### PR TITLE
Decay Function Improvements

### PR DESCRIPTION
**Pull request by BrandonGarciaWx**
- Changed names of decay functions to specify which should be used for learning rate and which should be used for sigma. This avoids confusion where a decay function meant for the learning rate could be used by users to decay sigma and thus potentially lead to overfitting.
- Replaced default sigma value with the hypotenuse of the SOM as this is a common starting sigma value that helps SOMs converge on global minimum. This avoids users picking small sigmas that erroneously converge on local quantization error minima.
- Replaced the asymptotic_decay function with the sigma_inverse_time_decay function. This new function always decays to a value of 1 to avoid overfitting. The asymptotic_decay function struggles to converge on solutions for sigma values greater than 3 due to the sigma at the end of the training being larger than 1 which results in less fine-scale tuning that usually occurs when the sigma value is near a value of 1.
- Updated test decay functions.
- Updated descriptions of decay functions and MiniSom class parameters.

These changes were previously discussed in #184

### CODE REVIEW

**Comment by JustGlowing** [comment]
hi @BrandonGarciaWx , thanks for raising this PR. I'll review it in the following days.

In the meanwhile, can you point out a reference to justify that the hypotenuse of the map is a good initial value for sigma?

File: minisom.py [hunk 1]
```diff
@@ -71,8 +71,14 @@ def fast_norm(x):
     return sqrt(dot(x, x.T))
 
 
-def asymptotic_decay(learning_rate, t, max_iter):
```
Comment (JustGlowing) [inline minisom.py:74]: Please don't delete this function, users of past versions might still want to use it.

File: minisom.py [hunk 2]
```diff
@@ -71,8 +71,14 @@ def fast_norm(x):
...
-    """Decay function of the learning process.
+def learning_rate_inverse_time_decay(learning_rate, t, max_iter):
+    """Decay function of the learning process that asymptotically
+    approaches zero.
+
+    This function should NOT be used for the sigma_decay_function
```
Comment (JustGlowing) [inline minisom.py:78]: If these is not supposed to be used for sigma, it's better to have a close selection just as the neighborhood function. This how the neighborhood function is selected: https://github.com/JustGlowing/minisom/blob/master/minisom.py#L210

File: minisom.py [hunk 3]
```diff
@@ -84,26 +90,58 @@ def asymptotic_decay(learning_rate, t, max_iter):
...
+    C = (sigma - 1) / max_iter
+    return sigma / (1 + (t * C))
 
 
 class MiniSom(object):
     Y_HEX_CONV_FACTOR = (3.0 / 2.0) / sqrt(3)
 
-    def __init__(self, x, y, input_len, sigma=1.0, learning_rate=0.5,
-                 decay_function=asymptotic_decay,
+    def __init__(self, x, y, input_len, sigma='hypotenuse', learning_rate=0.5,
```
Comment (JustGlowing) [inline minisom.py:140]: I'd prefer sigma not to accept a string, but have None as default.

**Comment by BrandonGarciaWx** [comment]
Thank you for the comments, I believe I have addressed them all. I also made some additional minor changes to some decay function descriptions and names.

My justification for the hypotenuse of the SOM being a good initial value for sigma comes from the following paper:
[Gervais_etal_2016.pdf](https://github.com/JustGlowing/minisom/files/15019019/Gervais_etal_2016.pdf). Specifically, the sentence on page 2361 that reads 
> It is good practice to begin the training with a radius of influence equal to the diameter of the SOM to ensure that all nodes are updated and decrease the value with training time (Kohonen 2001).
I know the lead author personally and confirmed that "diameter" and "hypotenuse" mean the same thing in this context.

The citation for Kohonen (2001) is 
> Kohonen, 2001: Self-Organizing Maps. 3rd ed. Springer Series in Information Sciences, Vol. 30, Springer, 502 pp.

File: minisom.py [hunk 4]
```diff
@@ -170,11 +230,31 @@ def euclidean(x, w):
...
 
-        sigma_decay_function : function (default=asymptotic_decay)
+        sigma_decay_function : string, optional
+        (default='inverse_decay_to_one')
             Function that reduces sigma at each iteration.
+
+            The default function is:
+                sigma(t) = sigma / (1 + (t * (sigma - 1) / max_iter))
+
+            A custom decay function will need to to take in input
```
Comment (JustGlowing) [inline minisom.py:240]: This is not true anymore

File: minisom.py [hunk 5]
```diff
@@ -170,11 +230,31 @@ def euclidean(x, w):
...
+            three parameters in the following order:
+
+            1. Sigma
+            2. Current iteration
+            3. Maximum number of iterations allowed
+
+            To prevent overfitting, custom decay functions should not
```
[the remark this answers is not in the document]
Comment (BrandonGarciaWx) [reply minisom.py:247]: I removed this text since custom decay functions are no longer supported/necessary.

**Comment by BrandonGarciaWx** [comment]
In addition to removing the text about custom decay functions, I added a function for sigma to linearly decrease to one.

File: minisom.py [hunk 6]
```diff
@@ -71,39 +71,110 @@ def fast_norm(x):
...
+    """
+    return sigma + (t * (1 - sigma) / max_iter)
+
+
+def asymptotic_decay(dynamic_parameter, t, max_iter):
+    """Legacy default decay function of the learning process
+    and sigma that decays these values asymptotically to 1/3
+    of their original values.
+
+    Using this function may lead to overfitting for sigma values
```
Comment (JustGlowing) [inline minisom.py:152]: I understand that is motivating your changes but I don't agree with this for the following reasons:
- Sigma has a different meaning according to the neighborhood function used.
- Even if you consider a neighborhood gaussian function, a sigma below zero at the end of the training doesn't necessarily results in overfitting if the learning rate decreases appropriately.

**Comment by JustGlowing** [comment]
> In addition to removing the text about custom decay functions, I added a function for sigma to linearly decrease to one.

See my comment here: https://github.com/JustGlowing/minisom/pull/185/files#r1572485322

File: minisom.py [hunk 7]
```diff
@@ -71,39 +71,110 @@ def fast_norm(x):
...
+    t : int
+        Current iteration.
+
+    max_iter : int
+        Maximum number of iterations for the training.
+    """
+    return learning_rate * (1 - t / max_iter)
+
+
+def inverse_decay_to_one(sigma, t, max_iter):
```
Comment (JustGlowing) [inline minisom.py:111]: There's no need anymore for the decay functions to be here. They're now better suited as methods just as the neighborhood functions.

File: minisom.py [hunk 8]
```diff
@@ -170,11 +238,17 @@ def euclidean(x, w):
         random_seed : int, optional (default=None)
             Random seed to use.
 
-        sigma_decay_function : function (default=asymptotic_decay)
+        sigma_decay_function : string, optional
```
Comment (JustGlowing) [inline minisom.py:241]: All the possible values should be listed in the comment. Like here: https://github.com/JustGlowing/minisom/blob/master/minisom.py#L155

File: minisom.py (hunk 6 above)
Comment (BrandonGarciaWx) [reply minisom.py:152]: Good points. Regardless of the neighborhood function used, a sigma value near zero would be updating few surrounding neurons. This could lead to overly localized training where the SOM is training off of what is essentially noise as opposed to the broader trends. However, I agree that there are cases where the learning rate would decrease enough so that this is not an issue. For simplicity, I removed this text along with the corresponding warning message to leave it up to the user to decide whether their SOM is training well.

**Comment by BrandonGarciaWx** [comment]
Decay functions have been moved and function descriptions have been updated.

**Comment by JustGlowing** [comment]
Thanks for your contribution @BrandonGarciaWx !
